runtime/domain.c

```diff
@@ -1325,19 +1362,44 @@ static void decrement_stw_domains_still_processing(void)
   }
 }
 
-static void stw_handler(caml_domain_state* domain)
+/* Wait for other running domains to stop */
+static void stw_wait_for_running(caml_domain_state* domain)
 {
-  CAML_EV_BEGIN(EV_STW_HANDLER);
-  CAML_EV_BEGIN(EV_STW_API_BARRIER);
-  {
-    SPIN_WAIT {
-      if (atomic_load_acquire(&stw_request.domains_still_running) == 0)
+  /* The STW leader issues interrupts to all threads, then checks if
+     all threads have been successfully interrupted, before flipping
+     the barrier we are waiting on; this tends to (and should) be
+     fast, but we likely need to wait a bit in any case */
+
+  if (stw_request.enter_spin_callback) {
+    /* Spin while there is useful work to do */
+    SPIN_WAIT_BOUNDED {
+      if (!atomic_load_acquire(&stw_request.domains_still_running.value)) {
+        return;
+      }
+
+      if (!stw_request.enter_spin_callback
+            (domain, stw_request.enter_spin_data)) {
         break;
+      }
+    }
+  }
 
-      if (stw_request.enter_spin_callback)
-        stw_request.enter_spin_callback(domain, stw_request.enter_spin_data);
+  /* Spin a bit for the other domains */
+  SPIN_WAIT_BOUNDED {
+    if (!atomic_load_acquire(&stw_request.domains_still_running.value)) {
```

Comment:
`Barrier_released`, I think (it might be interesting to test this code by changing the values of the `Barrier_` constants, in particular by trying a variation in which a different one of them is zero).